Validation to error out on using resrved keywords as fieldnames

indradeep opened the pull request:
Created a new constraint for YAML validation to be able to specify exclusion criteria for identifiers.

Added the set of reserved keywords in the streaming and window DTC validation schema to error out on reserved keyword fieldnames.

Fixes #86

mserranom commented on tests/core/syntax/schema_validator_test.py:
```diff
@@ -86,3 +86,10 @@ def test_invalid_datagroup_has_no_fields():
     assert 'DataGroups.0.Fields: Required field missing' in str(err.value)
+
+
+def test_invalid_field_name_reserved_keywords():
```
as in https://github.com/productml/blurr/pull/133:

my rule of thumb to have proper test case names: it should have at least a __verb__, otherwise it can't be a spec. For this case, something like:

This could be: `test_reserved_keywords_as_field_names_raises_invalid_schema_error`

indradeep replied:
Incorporated change.

prashastk commented on blurr/core/syntax/dtc_streaming_schema.yml:
```diff
@@ -33,7 +33,7 @@ data_group_block:
 
 field:
-  Name: identifier()
+  Name: identifier(exclude=['name', 'identity', 'needs_evaluation', 'schema', 'snapshot', 'nested_items', 'evaluation_context'])
```
I thought we decided to use _name, etc for local variables? Then we just need to ensure that no identified should start with _ as they are supposed to be for internal use.

indradeep commented:
As per discussion with @prashastk I have renamed all the local fields to start with _ so that they do not conflict with the user-defined fields.  Also, changed the validator to error out on user defined field names that start with _

prashastk commented on blurr/core/block_data_group.py:
```diff
@@ -37,8 +37,7 @@ def extend_schema(self, spec: Dict[str, Any]) -> Dict[str, Any]:
+    def _build_time_fields_spec(name_in_context: str) -> List[Dict[str, Any]]:
         """
         Constructs the spec for predefined fields that are to be included in the master spec prior to schema load
         :param name_in_context: Name of the current object in the context
```
Should we change the name of the predefined fields as well and make them start with a _? e.g. _start_time?
What happens if someone creates a start_time field?

indradeep replied:
Agreed.  Changing

prashastk commented on blurr/core/syntax/dtc_window_schema.yml:
```diff
@@ -32,8 +32,8 @@ data_group_window:
 
 field:
-  Name: identifier()
+  Name: identifier(exclude=['name', 'identity', 'needs_evaluation', 'schema', 'snapshot', 'nested_items', 'evaluation_context', 'window_source'])
```
Do we still need needs_evaluation etc here?

indradeep replied:
Nope, we no longer need these.  Cleaning up.

prashastk commented on blurr/core/window_data_group.py:
```diff
@@ -56,28 +56,28 @@ def prepare_window(self, start_time: datetime) -> None:
-        if self.schema.window_type == 'day' or self.schema.window_type == 'hour':
+        store = self._schema.source.store
+        if self._schema.window_type == 'day' or self._schema.window_type == 'hour':
             self.window_source.view = self._load_blocks(
```
Should this be _window_source as well?

indradeep replied:
Yes

prashastk commented on blurr/core/syntax/schema_validator.py:
```diff
@@ -41,23 +42,51 @@ def get_name(self) -> str:
     TAG = 'expression'
 
+    set_not_allowed = '\'%s\' sets value using `=`.'
+    invalid_python_expression = '\'%s\' is an invalid python expression.'
```
Make these two as CAPS as they are pretty much constants here.

indradeep replied:
Done.

prashastk commented on tests/core/syntax/schema_validator_test.py:
```diff
@@ -78,11 +95,20 @@ def test_invalid_incorrect_expression():
+
+
+def test_reserved_keywords_as_field_names_raises_invalid_schema_error():
+    with raises(InvalidSchemaError) as err:
```
You can add a match='Name: \'_name\' starts with _ or containing whitespace characters.' here instead of the assert below.

prashastk replied:
Same for other instances in this file.

indradeep replied:
Done.
